Add new `reindex-management` module

szybia opened the pull request:
- Add new `reindex-management` module which will primarily contain new endpoints for reindex resilience
  * extends reindex to get access to FeatureFlag
  * all functionality gated behind FeatureFlag so no changes in release builds
  * new package `org.elasticsearch.reindex.management`
  * defines new NodeFeature to-be-used by endpoints to guard multi-cluster calls where some nodes might not support new endpoints
  * i've tested rebasing my cancel impl on top of this and works swell

szybia commented on modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexPlugin.java:
```diff
@@ -45,11 +44,6 @@ public class ReindexPlugin extends Plugin implements ActionPlugin {
-    /**
-     * Whether the feature flag to guard the work to make reindex more resilient while it is under development.
-     */
-    static boolean REINDEX_RESILIENCE_ENABLED = new FeatureFlag("reindex_resilience").isEnabled();
```
not referenced by anything, and FeatureFlag constructor doesn't seem to have any side-effects

szybia commented on modules/reindex-management/src/main/java/org/elasticsearch/reindex/management/ReindexManagementFeatures.java:
```diff
@@ -0,0 +1,25 @@
+ * License v 3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.reindex.management;
```
makes the most sense to me

szybia committed:
Add new `reindex-management` module

szybia committed:
make ReindexPlugin extensible

szybia commented on modules/reindex-management/src/main/java/org/elasticsearch/reindex/management/ReindexManagementFeatures.java:
```diff
@@ -0,0 +1,26 @@
+
+public class ReindexManagementFeatures implements FeatureSpecification {
+
+    public static final NodeFeature NEW_ENDPOINTS = new NodeFeature("reindex_management_endpoints");
```
wondering do i just make this a `reindex_resilience`-wide feature, for relocating tasks later too

samxbr replied:
+1 for `reindex_resilience`, I doubt we will release the reindex API's (we could, but not much value) without the relocation piece.

samxbr commented on modules/reindex-management/src/main/java/org/elasticsearch/reindex/management/ReindexManagementFeatures.java:
```diff
@@ -0,0 +1,26 @@
+
+public class ReindexManagementFeatures implements FeatureSpecification {
+
+    public static final NodeFeature NEW_ENDPOINTS = new NodeFeature("reindex_management_endpoints");
```
My knowledge on plugins is lacking, is the new `NodeFeature` only accessible to test clusters in `reindex-management` or is it possible to load the `reindex-management` in test clusters for mixed cluster tests in `reindex` module as well?

I am imagining scenarios for testing PIT reindex, where the production code probably makes more sense to live in `reindex` instead of `reindex-management`, but the tests will probably require the `NodeFeature` in a mixed cluster set up. Or maybe those tests will need to be moved to `reindex-management` as well, though it seems a bit weird?

samxbr commented on modules/reindex-management/src/main/resources/META-INF/services/org.elasticsearch.features.FeatureSpecification:
```diff
@@ -0,0 +1,10 @@
+ # License v3.0 only", or the "Server Side Public License, v 1".
+#
+
+org.elasticsearch.reindex.management.ReindexManagementFeatures
```
Do you need to add the class to `module-info.java` (similar to [this](https://github.com/elastic/elasticsearch/blob/main/modules/data-streams/src/main/java/module-info.java))?

PeteGillinElastic reviewed (approved):
LGTM, thanks.

Not sure whether we should wait for Ryan to look or not...

samxbr reviewed (approved):
LGTM, +1 on nothing under discussion needs to block this PR

szybia committed:
fix `ml/qa/native-multi-node-tests` failures

my understanding: these tests add a new node that must have the same NodeFeature
definitions as the `testClusters` cluster started by gradle in order to
be able to join the cluster.

therefore we need to drag it in here too so the NodeFeatures exist on
new nodes.

i've checked other PR which added `javaRestTestImplementation` to this
file and they all seem to introduce new NodeFeatures and a matching entry in this
file, similar to me:
1. https://github.com/elastic/elasticsearch/pull/104466/changes#diff-c7c94cdf3a65fbfe4a3f736a490eccef193ad9aa29c6f5e4e5c125a6049048faR18
2. https://github.com/elastic/elasticsearch/pull/114168/changes#diff-c7c94cdf3a65fbfe4a3f736a490eccef193ad9aa29c6f5e4e5c125a6049048faR21
3. https://github.com/elastic/elasticsearch/pull/112348/changes#diff-c7c94cdf3a65fbfe4a3f736a490eccef193ad9aa29c6f5e4e5c125a6049048faR20

szybia committed:
Merge branch 'main' into reindex-mgmt-new-module
